In [11]:
from IPython.display import display

import sympy
import scipy
import numpy as np

# x, y, z = symbols('x y z')
sympy.init_printing(use_unicode=True)

# Homework

## Exercise set 5.1

### 4a

For each choice of $f(t, y)$ given in parts (a):

**i.** Does $f$ satisfy a Lipschitz condition on $D={(t, y) | 0 \le t \le 1, -\infty < y < \infty }$?

**ii.** Can Theorem 5.6 be used to show that the initial-value-problem
$$y'=f(t, y), 0 \le t \le 1, y(0)=1,$$
is well-posed?

**a.** $f(t, y) = e^{t-y}$

**solution**

$f(t, y) = e^{t-y}$

$\frac{d}{dy}f(t, \xi)=-e^{t-\xi}$

$|f(t, y_2)-f(t, y_1)| = |y_2 - y_1||-e^{t-\xi}| \le -|y_2 - y-_1|$

**i.** Yes, $L=-1$

**ii.** Yes

### 4c

**c.** $f(t, y) = \cos(yt)$

**solution**

$\frac{d}{dy}f(t, \xi)=-t \sin(t \xi)$

$|f(t, y_2)-f(t, y_1)| = |y_2 - y_1||-t \sin(t \xi)| \le |y_2 - y-_1|$

**i.** Yes, $L=1$

**ii.** Yes

## Exercise set 5.2

### 2c

Use Euler's method to approximate the solutions for each of the following initial-value problems.

**c.** $y'=-y + ty^{1/2}, 2 \le t \le 3, y(2)=2$, with $h=0.25$.

**solution**

```
t_i  w_i
____________
2.00 2.00000
2.25 2.20711
2.50 2.49100
2.75 2.85468
3.00 3.30260
```

In [12]:
def euler(f, a, b, N, alpha):
    out = []
    h = (b - a)/N
    t = a
    w = alpha
    out.append((t, w))
    for i in range(1, N+1):
        w = w + h * f(t, w)
        t = a + i * h
        out.append((t, w))
    return out

# Sanity check example from text
# f = lambda t, y: y - np.power(t, 2) + 1
# euler(f, 0, 2, 10, 0.5)

print("t_i  w_i")
print("_"*12)
f = lambda t, y: -1 * y + t * np.sqrt(y)
for t, w in euler(f, 2, 3, 4, 2):
    print(f"{t:.2f} {round(w, 5):.5f}")


t_i  w_i
____________
2.00 2.00000
2.25 2.20711
2.50 2.49100
2.75 2.85468
3.00 3.30260


### 4c

The acutal solutions to the initial-value problems in Exercise 2 are given here.
Compute the actual error and compare this to the error bound if Theorem 5.9 can be applied.

**c.** $y(t)=\big(t-2+\sqrt{2} e e^{-t/2}\big)^2$

**solution**

&emsp; $t_i$ &emsp; $w_i$ &emsp; $y_i=y(t_i)$ &emsp; $|y_i - w_i|$

```
2.00 2.00000 2.0 0.0
2.25 2.20711 2.24412 0.03701
2.50 2.49100 2.56445 0.07345
2.75 2.85468 2.96519 0.11051
3.00 3.30260 3.45129 0.14869
```

In [13]:
y_t = lambda t: np.power(t - 2 + np.sqrt(2) * np.e * np.power(np.e, -1 * t / 2), 2)
real = []
for i in range(5):
    t = i * 1/4 + 2
    real.append(y_t(t))

f = lambda t, y: -1 * y + t * np.sqrt(y)
approx = euler(f, 2, 3, 4, 2)
errs = []
for i in range(5):
    errs.append(abs(real[i] - approx[i][1]))

print("t_i  w_i y_i=y(t_i) |y_i - w_i|")
print("_"*20)
for i in range(5):
    print(f"{approx[i][0]:.2f} {round(approx[i][1], 5):.5f} {round(real[i], 5)} {round(errs[i], 5)}")

t_i  w_i y_i=y(t_i) |y_i - w_i|
____________________
2.00 2.00000 2.0 0.0
2.25 2.20711 2.24412 0.03701
2.50 2.49100 2.56445 0.07345
2.75 2.85468 2.96519 0.11051
3.00 3.30260 3.45129 0.14869


## Exercise set 5.3

### 10

Given the initial-value problem

$$y'=\frac{1}{t^2}-\frac{y}{t} - y^2, 1 \le t \le 2, y(1)=-1$$

with exact solution $y(t)=-1/t$:

**a.** Use Taylor's method of order two with h=0.05 to approximate the solution, and compare it with the acutal values of $y$.

**solution**

$$f(t, y(t)) = \frac{1}{t^2}-\frac{y(t)}{t} - y(t)^2$$

$$f'(t, y(t)) = \frac{d}{dt}\big(\frac{1}{t^2}-\frac{y}{t} - y^2\big) = \frac{y'}{t^2} - \frac{2}{t^3}$$

$$T^{(2)}(t_i, w_i) = f(t_i, w_i) + \frac{h}{2}f'(t_i, w_i) = \frac{1}{t_i^2}-\frac{w_i}{t_i} - w_i^2 + \frac{h}{2} \bigg( \frac{w_i}{t_i^2} - \frac{2}{t_i^3} \bigg)$$

In [14]:
# t, y = sympy.symbols('t y')
# yp = 1/t**2 - y/t - y**2
# sympy.diff(yp, t)

In [15]:
y = lambda t: -1 / t
yp = lambda t, w: 1/np.power(t, 2) - w/t - np.power(w, 2)

def taylor2(t_i, w_i):
    h = 0.05
    return yp(t_i, w_i) + h/2*(-2/np.power(t_i, 3) + yp(t_i, w_i)/np.power(t_i, 2) - np.power(yp(t_i, w_i), 2))

w = -1
for n in range(1, 21):
    i = 0.05 * n
    t = 1 + i
    w = taylor2(t, w)
    print(t, w)
    # print(f'{t:.2f}, {round(w, 5):.5f}, {round(y(t), 5):.5f}, {round(abs(y(t) - w), 5)}')

1.05 0.8172416585681892
1.1 -0.6425620463796816
1.15 0.8658422265152941
1.2 -0.8342787688307512
1.25 0.5862390735175146
1.3 -0.2297027391676198
1.35 0.6438068321489563
1.4 -0.39032650448017464
1.45 0.5743288174244908
1.5 -0.28789045050627665
1.55 0.5043263116610208
1.6 -0.19367865380631955
1.65 0.43515554987500493
1.7000000000000002 -0.11059621726629576
1.75 0.3676864906606306
1.8 -0.03965661164906721
1.85 0.30399561073974446
1.9 0.017462812333395618
1.9500000000000002 0.24704017693440036
2.0 0.05950303561412426


**b.** Use the answers generated in part (a) and linear interpolation to approximate the following values of $y$, and compare them to the actual values.

&emsp; **i.** $y(1.052)$ &emsp; **ii.** $y(1.555)$ &emsp; **iii.** $y(1.978)$

**solution**

**c.** Use Taylor's method of order four with h=0.05 to approximate the solution, and compare it with the acutal values of $y$.

**solution**

**d.** Use the answers generated in part (c) and piecewise cubic Hermite interpolation to approximate the following values of $y$, and compare them to the actual values.

&emsp; **i.** $y(1.052)$ &emsp; **ii.** $y(1.555)$ &emsp; **iii.** $y(1.978)$

**solution**

In [16]:
# Hermite interpolation
x0 = -0.25
x1 = 0.25
data = np.array([
    [x0**3, x0**2, x0, 1],
    [x1**3, x1**2, x1, 1],
    [3*(x0**2), 2*x0, 1, 0],
    [3*(x1**2), 2*x1, 1, 0],
])
dep = np.array([1.33203, 0.800781, 0.437500, -0.625000])
np.linalg.solve(data, dep)

array([ 7.749984  , -1.0625    , -1.546872  ,  1.13281175])

## Exercise set 5.4

### 4a

Use the Modified Euler method to approximate the solutions to each of the folowing initial-value problems, and compare the results to the actual values.

**a.** $y'=\frac{2-2ty}{t^2+1}, 0 \le t \le 1, y(0)=1$, with $h=0.1$; actual solution $y(t)=\frac{2t+1}{t^2+1}$

**solution**

### 8a

Repeat Exercise 4 using the Midpoint method.

**solution**

### 12a

Repeat Exercise 4 using Heun's method.

**solution**

### 16a

Repeat Exercise 4 using the Runge-Kutta method of order four.

**solution**

### 26a

Compare the results of Exercise 16 and Cubic Hermite interpolation to approxiamte the values of $y(t)$, and compare the approximations to the actual values.

**a.** $y(1.25)$ and $y(1.93)$

**solution**

## Exercise set 5.5

### 2b

Use the Runge-Kutta-Fehlberg Algorithm with tolerance $TOL=10^{-4}$ to approximate the solution to the following initial-value problems.

**b.** $y'=\sin t + e^{-t}, 0 \le t \le 1, y(0) = 0$, with $hmax=0.25$ and $hmin=0.02$.

**solution**